File: Arrays/threesum_15.py
# -*- coding: utf-8 -*-
"""
Created on Wed Jul 13 21:52:01 2022

@author: yuxin_liu_1998
"""


# with two sum as basic algorithm
def threeSum( nums: list[int]) -> list[list[int]]:
        # sort arrays
        nums = sorted(nums)
        res = set()
        
        # iterate through nums
        for i in range(len(nums)):
            if nums[i] > 0:
                break
            target = -nums[i]
            # using two sum algotithm to solve for each target
            value_dict = dict()
            for j in range(i+1,len(nums)):
                if target-nums[j] in value_dict:
                    temp_res = (nums[i],target-nums[j],nums[j])
                    if temp_res not in res:
                        res.add(temp_res)
                else:
                    value_dict[nums[j]] = value_dict.get(nums[j],0) + 1
        # 
        out = []
        for item in res:
            out.append(list(item))
        
        return out


# solution 2 double pointer
def threeSum( nums: list[int]) -> list[list[int]]:
        # sort arrays
        nums = sorted(nums)
        res = set()
        n = len(nums)
        i = 0
        
        while True:
            if i>=n or nums[i] > 0:
                break
            l = i+1
            r = n-1      
            while True:
                if l >= r:
                    break
                total = nums[i]+nums[l]+nums[r]
                if total == 0:
                    res.add((nums[i],nums[l],nums[r]))
                    l += 1
                    r -= 1
                elif total > 0:
                    r -= 1
                else:
                    l += 1
            i += 1
        out = list([list(item) for item in res])           
        return out



nums = [-1,0,1,2,-1,-4]
nums = [0,1,1]
nums = [0,0,0]
nums = [3,0,-2,-1,1,2]
out = threeSum(nums)

File: hashTable/ThreeSum15.java
import java.util.*;

public class ThreeSum15 {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        Set<Set<Integer>> tempres = new HashSet<>();
        Arrays.sort(nums);
        int n = nums.length;
        // two pointer
        for (int i = 0; i < n-2; i++) {
            if (nums[i] >0 ) {
                return res;
            }
            if(i>0 && nums[i]==nums[i-1]){
                continue;
            }
            int left = i+1;
            int right = n-1;

            while(right>left){
                int sum = nums[i]+nums[left]+nums[right];
                if(sum==0){
                    res.add(Arrays.asList(nums[i], nums[left], nums[right]));
                    // remove duplicate, b  c
                    while (right > left && nums[right] == nums[right - 1]) right--;
                    while (right > left && nums[left] == nums[left + 1]) left++;
                    right--;
                    left++;
                } else if(sum>0){
                    right --;
                }else{
                    left ++;
                }

            }
        }
        return res;

    }


}
